Added docker support, part 2

Duvel opened the pull request:
This pull request is based on the work of @jezdez in #456, also a mention to msgre/uwsgi_logging@4055e52 for the extra logging.

One now can run `docker build --tag umap:1.2.2 .` and this will result in a working image for uMap.

Paris999 commented:
I have tried to run the command `docker build --tag umap:1.2.2 .` but I got only the Error `unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /home/sysop/Dockerfile: no such file or directory`

What I have to do, that it is working.
thanks.

Paris999 commented:
Okay now I cloned with git the forked project. and it was starting, in the End it show successfully created.
But during the installtion I saw two Error
`npm install

> phantomjs@1.9.7-15 install /srv/umap/node_modules/mocha-phantomjs/node_modules/phantomjs
> node install.js

invalid configloglevel="notice"
Downloading https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-1.9.7-linux-x86_64.tar.bz2
Saving to /srv/umap/node_modules/mocha-phantomjs/node_modules/phantomjs/phantomjs/phantomjs-1.9.7-linux-x86_64.tar.bz2
Receiving...`

and

`Get:159 http://deb.debian.org/debian buster/main amd64 uwsgi amd64 2.0.18-1 [51.6 kB]
debconf: delaying package configuration, since apt-utils is not installed
Fetched 124 MB in 12s (10.6 MB/s)
Selecting previously unselected package perl-modules-5.28.
(Reading database ... 6829 files and directories currently installed.)`

after it was created, I started it with run under docker und than I tried to start it, but it is exiting direct again. I don't know if it has something to do with the to errors during the installation.
any ideas?

Duvel commented:
I think those errors are normal during the build, I didn't even notice them I guess. You need to do more to get the image running then just to run it. You would also need a database and a redis server.

This is the docker stack file, that I use to run it:
```

version: '3'
services:
  db:
    image: mdillon/postgis:9.6-alpine
    volumes:
      - db:/var/lib/postgresql/data

  redis:
    image: redis:latest

  app:
    image: umap:1.2.2
    environment:
      - DATABASE_URL=postgis://postgres@db/postgres
      - REDIS_URL=redis://redis:6379/0
      - SECRET_KEY=some-long-and-weirdly-unrandom-secret-key
      - ALLOWED_HOSTS=*
      - SITE_URL=https://umap.somewhere.nl/
      - LEAFLET_STORAGE_ALLOW_ANONYMOUS=True
    volumes:
      - SOMEDIR/uploads:/srv/umap/uploads


volumes:
  db:

```

Loki-Afro commented:
whats the status here? I'd like to have a dockerized umap too :)

0x6b-dev commented:
```mkdir: cannot create directory '/usr/share/man/man1': File exists
The command '/bin/sh -c mkdir /usr/share/man/man1' returned a non-zero code: 1```

I'm getting the above error message trying to build the container.

Please merge proper dockerfiles in.

MaxBarbet commented:
> ```
> mkdir: cannot create directory '/usr/share/man/man1': File exists
> The command '/bin/sh -c mkdir /usr/share/man/man1' returned a non-zero code: 1
> ```
> 
> I'm getting the above error message trying to build the container.
> 
> Please merge proper dockerfiles in.

I have the same problem here too

micheljung commented:
@Duvel are you still around? Would you mind rebasing?

Duvel commented:
@micheljung I merged the latest changes and was able to create a build. But a lot more has to probably to be done.

ALLOWED_HOSTS is now a list and apparently you need at least postgres 11 for it to run.

Duvel commented:
With this docker-compose file and the latest changes from my repo, I have something starting again, but it gives an error.

```
version: '3'
services:
  db:
    image: duvel/postgis:12-2.5
    environment:
      - POSTGRES_HOST_AUTH_METHOD=trust
    volumes:
      - db:/var/lib/postgresql/data

  redis:
    image: redis:latest

  app:
    image: umap:1.2.3
    ports:
      - "8001:8000"
    environment:
      - DATABASE_URL=postgis://postgres@db/postgres
      - REDIS_URL=redis://redis:6379/0
      - SECRET_KEY=some-long-and-weirdly-unrandom-secret-key
      - ALLOWED_HOSTS=[*]
      - SITE_URL=https://umap.somewhere.nl/
      - LEAFLET_STORAGE_ALLOW_ANONYMOUS=True
    volumes:
      - SOMEDIR/uploads:/srv/umap/uploads

volumes:
  db:
```

I don't know how to solve this error, it seems to do with Django 4.

```
app_1    | django.template.library.InvalidTemplateLibrary: Invalid template library specified. ImportError raised when trying to load 'compressor.templatetags.compress': cannot import name 'force_text' from 'django.utils.encoding' (/usr/local/lib/python3.8/site-packages/django/utils/encoding.py)
```

pommi commented:
Hi all, @Duvel thanks for all the work you've done in this PR.

I wanted to run a uMap instance in a container, but I got stuck with what you have in this PR so far. I took your contribution as an inspiration and built something similar.

I took a little different approach and built the container with the [installation tutorial](https://umap-project.readthedocs.io/en/latest/ubuntu/) in mind, configuring everything as described in the installation tutorial and taking the latest version (currently 1.2.3) from PyPI. While I see quite some `Exception while resolving variable 'x' in template 'y'.` lines in the logs, basic functionality works (for me).

If you want to quickly run uMap in a container, feel free to try it out: https://hub.docker.com/repository/docker/pommib/umap

tuxrinku commented:
@Duvel I managed to get the container working by simply replacing the setup.cfg file by the one in the umap repo. Basically it just prevents django >= 4 from being installed, and no errors are thrown. Thank you for your work.

Duvel commented:
@tuxrinku ah, they solved the Django 4 issue a few days ago. I updated my fork so it is included, but now they have added an error in the Makefile, so the Docker build now fails on that.

Duvel commented:
@tuxrinku I solved the issue in the Makefile, had to do with a difference with calling cp.

Duvel commented:
@pommi nice work, I tried that procedure once, but more work was required then, but I guess with work from this PR it now can be done.

Duvel commented:
I updated the Dockerfile to nodejs 18 LTS and removed phantomjs. I also installed tini with the package manager.

It now also can be build for arm64, so  it can run on Raspberry Pi and Apple Silicon (M1).

xerz-one commented:
Question: shouldn't `createsuperuser` be added to the Dockerfile?

yohanboniface commented on Dockerfile:
```diff
@@ -0,0 +1,67 @@
+
+RUN make vendors
+
+FROM python:3.8-slim
```
Maybe 3.11 now ?

Duvel replied:
I'm not a Python programmer, so if that is the latest we can use that :)

yohanboniface commented on Dockerfile:
```diff
@@ -0,0 +1,67 @@
+
+COPY . /srv/umap
+
+COPY --from=vendors /srv/umap/umap/static/umap/vendors /srv/umap/umap/static/umap/vendors
```
This should not be needed if using a stable release instead of the sources

Duvel replied:
Maybe @pommi can share some of his work as he followed the steps in the installation.

yohanboniface commented on Makefile:
```diff
@@ -41,7 +41,8 @@ vendors:
 	mkdir -p umap/static/umap/vendors/togpx/ && cp -r node_modules/togpx/togpx.js umap/static/umap/vendors/togpx/
 	mkdir -p umap/static/umap/vendors/tokml && cp -r node_modules/tokml/tokml.js umap/static/umap/vendors/tokml
-	mkdir -p umap/static/umap/vendors/locatecontrol/ && cp -r node_modules/leaflet.locatecontrol/{dist/L.Control.Locate.css,src/L.Control.Locate.js} umap/static/umap/vendors/locatecontrol/
+	mkdir -p umap/static/umap/vendors/locatecontrol/ && cp -r node_modules/leaflet.locatecontrol/dist/L.Control.Locate.css umap/static/umap/vendors/locatecontrol/
```
Why this change ?

yohanboniface replied:
I'd say that's one more argument for not trying to build uMap from sources :)

yohanboniface reviewed (commented):
@Duvel Thanks for your work on this, and sorry for my very late answer.

I'd like this PR to be merge and to have and maintain an official uMap Docker image.

I've let a few comments for you.

Duvel commented on umap/settings/docker.py:
```diff
@@ -0,0 +1,155 @@
+# -*- coding:utf-8 -*-
+"""
```
With docker and kubernetes a lot of configuration can be done with environment variables and not with setting files. If I remember correctly this file and that dependency allows to do that.

Duvel replied:
An example of a configuration in docker can be found in this comment:
https://github.com/umap-project/umap/pull/801#issuecomment-1279765869

yohanboniface replied:
I think using env for reading some settings make sense, but:

- let's keep it only for the base settings (database URL, redis URL, secret link)
- this should be done in the `settings/base.py`, because maintaining a full settings for docker is not DRY at all and will make the docker image to break often I think

Duvel replied:
I would expect the basic settings to be set with environment variables and the more specific via other means.

Always supplying umap.conf doesn't really simplify things on docker or kubernetes. On kubernetes working with local files isn't recommended and would need extra setup and isn't possible on for instance Google Cloud. One could do it via config maps on kubernetes, but that is more difficult to setup.

Duvel commented on Dockerfile:
```diff
@@ -0,0 +1,67 @@
+FROM node:18 AS vendors
```
It is possible to have multiple stages in a Dockerfile and node isn't used here as the base image, but as a temporary stage to build the vendors, which is done with node apparently.

Duvel replied:
Yup, I think so to, the world has moved on since the first attempts at running umap  with docker.

yohanboniface commented:
Closing in favor of #1121 which is based on this PR. Hopefully we'll land it this time!
